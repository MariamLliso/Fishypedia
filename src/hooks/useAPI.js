import { endpoint } from "../paths/endpoint";
import { useCallback, useContext } from "react";
import FishesDataContext from "../store/context/FishesDataContext";
import {
  addLocalFishesActionCreator,
  deleteLocalFishesActionCreator,
  loadFishesActionCreator,
  loadLocalFishesActionCreator,
} from "../store/actions/actionCreators";

const useAPI = () => {
  const urlAPI = process.env.REACT_APP_API_URL;
  const localUrlAPI = process.env.REACT_APP_LOCAL_API_URL;
  const { dispatch, localDispatch } = useContext(FishesDataContext);

  const loadFishes = useCallback(async () => {
    const response = await fetch(`${urlAPI}${endpoint.fish}`);
    const dataFishes = await response.json();
    dispatch(loadFishesActionCreator(dataFishes));
    return dataFishes;
  }, [urlAPI, dispatch]);

  const loadLocalFishes = useCallback(async () => {
    const response = await fetch(`${localUrlAPI}${endpoint.fish}`);
    const dataFishes = await response.json();
    localDispatch(loadLocalFishesActionCreator(dataFishes));
    return dataFishes;
  }, [localDispatch, localUrlAPI]);

  const addToFishTank = async (newFish) => {
    const response = await fetch(`${localUrlAPI}${endpoint.fish}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(newFish),
    });
    if (response.ok) {
      const fish = await response.json();
      localDispatch(addLocalFishesActionCreator(fish));
      return fish;
    }
  };

  const deleteFishFromFishTank = async (id) => {
    const response = await fetch(`${localUrlAPI}${endpoint.fish}/${id}`, {
      method: "DELETE",
    });
    if (response.ok) {
      localDispatch(deleteLocalFishesActionCreator(id));
    }
  };

  return { loadFishes, loadLocalFishes, addToFishTank, deleteFishFromFishTank };
};

export default useAPI;
