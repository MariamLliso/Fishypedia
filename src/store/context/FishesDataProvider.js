import { useState, useReducer } from "react";
import fishesReducer from "../reducers/fishesReducer";
import localFishesReducer from "../reducers/localFishesReducer";
import FishesDataContext from "./FishesDataContext";

const FishesDataProvider = ({ children }) => {
  const [fishes, dispatch] = useReducer(fishesReducer, []);
  const [localFishes, localDispatch] = useReducer(localFishesReducer, []);
  const initialPage = "";
  const [page, setPage] = useState(initialPage);

  return (
    <FishesDataContext.Provider
      value={{ fishes, dispatch, localFishes, localDispatch, page, setPage }}
    >
      {children}
    </FishesDataContext.Provider>
  );
};

export default FishesDataProvider;
