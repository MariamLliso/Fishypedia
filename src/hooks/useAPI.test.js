import { renderHook, waitFor } from "@testing-library/react";
import FishesDataProvider from "../store/context/FishesDataProvider";
import useAPI from "./useAPI";
import TestRenderer from "react-test-renderer";

const expectFishes = [
  {
    id: 1,
    "file-name": "bitterling",
    name: {
      "name-USen": "bitterling",
      "name-EUen": "bitterling",
      "name-EUde": "Bitterling",
      "name-EUes": "amarguillo",
      "name-USes": "amarguillo",
      "name-EUfr": "bouvière",
      "name-USfr": "bouvière",
      "name-EUit": "rodeo",
      "name-EUnl": "bittervoorn",
      "name-CNzh": "红目鲫",
      "name-TWzh": "紅目鯽",
      "name-JPja": "タナゴ",
      "name-KRko": "납줄개",
      "name-EUru": "горчак",
    },
    availability: {
      "month-northern": "11-3",
      "month-southern": "5-9",
      time: "",
      isAllDay: true,
      isAllYear: false,
      location: "River",
      rarity: "Common",
      "month-array-northern": [11, 12, 1, 2, 3],
      "month-array-southern": [5, 6, 7, 8, 9],
      "time-array": [
        0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
        20, 21, 22, 23,
      ],
    },
    shadow: "Smallest (1)",
    price: 900,
    "price-cj": 1350,
    "catch-phrase": "I caught a bitterling! It's mad at me, but only a little.",
    "museum-phrase":
      "Bitterlings hide their eggs inside large bivalves—like clams—where the young can stay safe until grown. The bitterling isn't being sneaky. No, their young help keep the bivalve healthy by eating invading parasites! It's a wonderful bit of evolutionary deal making, don't you think? Each one keeping the other safe... Though eating parasites does not sound like a happy childhood... Is that why the fish is so bitter?",
    image_uri: "https://acnhapi.com/v1/images/fish/1",
    icon_uri: "https://acnhapi.com/v1/icons/fish/1",
  },
  {
    id: 2,
    "file-name": "pale_chub",
    name: {
      "name-USen": "pale chub",
      "name-EUen": "pale chub",
      "name-EUde": "Döbel",
      "name-EUes": "cacho",
      "name-USes": "cacho",
      "name-EUfr": "chevaine",
      "name-USfr": "chevaine",
      "name-EUit": "zacco",
      "name-EUnl": "Japanse drakenvis",
      "name-CNzh": "溪哥",
      "name-TWzh": "溪哥",
      "name-JPja": "オイカワ",
      "name-KRko": "피라미",
      "name-EUru": "светлый закко",
    },
    availability: {
      "month-northern": "",
      "month-southern": "",
      time: "9am - 4pm",
      isAllDay: false,
      isAllYear: true,
      location: "River",
      rarity: "Common",
      "month-array-northern": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
      "month-array-southern": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
      "time-array": [9, 10, 11, 12, 13, 14, 15],
    },
    shadow: "Smallest (1)",
    price: 200,
    "price-cj": 300,
    "catch-phrase": "I caught a pale chub! That name seems a bit judgy...",
    "museum-phrase":
      "The pale chub is a river fish with simple back-and-silver coloring. Interestingly, the males' coloring becomes most vibrant when he's trying to attract a mate! Though normally plain, these fellows really know how to look stylish when they want to. Perhaps I could learn a thing or two from the pale chub...",
    image_uri: "https://acnhapi.com/v1/images/fish/2",
    icon_uri: "https://acnhapi.com/v1/icons/fish/2",
  },
];

describe("Given the custom Hook useAPI", () => {
  describe("When loadFishes function it's called", () => {
    test("Then it should return an array of two fishes", async () => {
      const wrapper = ({ children }) => (
        <FishesDataProvider>{children}</FishesDataProvider>
      );

      const { result } = renderHook(() => useAPI(), { wrapper });
      const fishes = await waitFor(() => result.current.loadFishes());
      expect(fishes).toEqual(expectFishes);
    });
  });

  describe("When loadLocalFishes is called with an array of local fishes", () => {
    test("Then it should return an array with fishes", async () => {
      const wrapper = ({ children }) => (
        <FishesDataProvider>{children}</FishesDataProvider>
      );
      const { result } = renderHook(() => useAPI(), { wrapper });
      const localFishes = await waitFor(() => result.current.loadLocalFishes());
      expect(localFishes).toEqual(expectFishes);
    });
  });

  describe("When addToFishTank is called with a fish", () => {
    test("Then it should return the fish added", async () => {
      const expectFish = [
        {
          id: 29,
          "file-name": "golden_trout",
          name: {
            "name-USen": "golden trout",
            "name-EUen": "golden trout",
            "name-EUde": "Goldforelle",
            "name-EUes": "trucha dorada",
            "name-USes": "trucha dorada",
            "name-EUfr": "truite dorée",
            "name-USfr": "truite dorée",
            "name-EUit": "trota dorata",
            "name-EUnl": "goudforel",
            "name-CNzh": "金鳟",
            "name-TWzh": "金鱒",
            "name-JPja": "ゴールデントラウト",
            "name-KRko": "금송어",
            "name-EUru": "золотая форель",
          },
          availability: {
            "month-northern": "3-5 & 9-11",
            "month-southern": "3-5 & 9-11",
            time: "4pm - 9am",
            isAllDay: false,
            isAllYear: false,
            location: "River (Clifftop)",
            rarity: "Ultra-rare",
            "month-array-northern": [3, 4, 5, 9, 10, 11],
            "month-array-southern": [3, 4, 5, 9, 10, 11],
            "time-array": [
              16, 17, 18, 19, 20, 21, 22, 23, 0, 1, 2, 3, 4, 5, 6, 7, 8,
            ],
          },
          shadow: "Medium (3)",
          price: 15000,
          "price-cj": 22500,
          "catch-phrase":
            "I caught a golden trout! But the real treasure? Friendship.",
          "museum-phrase":
            "The golden trout is a beautifully colored fish that can only live in very clean waters. They are difficult to come across since they are found only in high mountain streams. As a side note, I find it much easier to appreciate fish that aren't such prima donnas about everything.",
          image_uri: "https://acnhapi.com/v1/images/fish/29",
          icon_uri: "https://acnhapi.com/v1/icons/fish/29",
        },
      ];

      const wrapper = ({ children }) => (
        <FishesDataProvider>{children}</FishesDataProvider>
      );
      const { result } = renderHook(() => useAPI(), { wrapper });
      const addToFishTank = await waitFor(() =>
        result.current.addToFishTank(expectFish)
      );
      expect(addToFishTank).toEqual(expectFish);
    });
  });
});

describe("Given useApi Hook", () => {
  describe("When rendered", () => {
    test("Then it should always match this snapshot", () => {
      const renderedUseApi = TestRenderer.create(
        <FishesDataProvider>
          <useAPI />
        </FishesDataProvider>
      );

      expect(renderedUseApi).toMatchSnapshot();
    });
  });
});
