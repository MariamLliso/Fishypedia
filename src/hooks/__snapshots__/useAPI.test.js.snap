// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given useApi Hook When rendered Then it should always match this snapshot 1`] = `<useAPI />`;
